cpp/include/ucxx/delayed_submission.h

```diff
@@ -19,13 +22,38 @@ namespace ucxx {
 
 typedef std::function<void()> DelayedSubmissionCallbackType;
 
+enum class DelayedSubmissionOperationType { Undefined = 0, AM, Stream, Tag, TagMulti };
+
+class DelayedSubmissionData {
+ public:
+  /**
+   * @brief Constructor for `DelayedSubmission` operation-specific data.
+   *
+   * Construct an object containing operation-specific data for `DelayedSubmission`.
+   *
+   * @param[in] operationType the type of operation the object refers.
+   * @param[in] memoryType    optional memoryType, only if operation type is AM.
+   * @param[in] tag           optional tag, only if operation type is Tag or TagMulti.
+   * @param[in] tagMask       optional tagMask, only if operation type is Tag or TagMulti.
+   */
+  explicit DelayedSubmissionData(const DelayedSubmissionOperationType operationType,
+                                 const std::optional<ucs_memory_type_t> memoryType,
+                                 const std::optional<ucp_tag_t> tag,
+                                 const std::optional<ucp_tag_t> tagMask);
+
+  const DelayedSubmissionOperationType _operationType{
+    DelayedSubmissionOperationType::Undefined};        ///< The operation type
+  const std::optional<ucs_memory_type_t> _memoryType;  ///< Memory type used on the operation
+  const std::optional<ucp_tag_t> _tag;                 ///< Tag to match
+  const std::optional<ucp_tag_t> _tagMask;             ///< Tag mask to use
+};
```

Comment:
**issue/question:**

I don't think this interface for wrapping up the packet of delayed submission data is very type safe (or memory efficient).

The `operationType` is, AIUI, a tag that says whether we're doing a tag send/recv or an active message.

If we're using active messages, then it is invalid to provide a tag and tagmask argument. But one _must_ provide a memory type. Conversely, for tag or tagmulti, one _must_ provide both a tag and a tagmask, but _must not_ provide memory type.

Further, if the operation type is `Undefined` then we don't have any information and the object is unusable.

If I understand things correctly, then the data is one of (I'm ignoring `Stream` for now, but I think pattern will still apply):

```
struct AM { ucs_memory_type_t mem_type; };
// If these two are the same for all purposes they don't need to be different structs.
struct Tag { ucp_tag_t tag; ucp_tag_t mask; };
struct TagMulti {ucp_tag_t tag; ucp_tag_t mask; };
```
And then this data is a `std::variant<AM, Tag, TagMulti> data`;

So you could imagine something like:
```c++
struct One {
  int x;
};
struct Two {
  int x;
  float y;
};

struct Data {
  std::variant<One, Two> var;
  Data(decltype(var) v) : var{v} {};

  One one() { return std::get<One>(var); }
  Two two() { return std::get<Two>(var); }
};

int main(void) {
  auto one = Data(One{1});
  auto two = Data(Two{1, 2});

  // throws std::bad_variant_access since two does not contain a One
  std::cout << two.one().x << std::endl;
  return 0;
}
```

The constructor of the data must know whether they are a tag/tagmulti/am or stream person so they can plonk the correct concrete struct in, and then they can also attempt to pull the correct variant out, and catch if the wrong type arrives. WDYT?

(It's not as ergonomic as Rust enums + pattern matching :disappointed: )

Reply:
Another great suggestion, in fact I initially tried getting something like that with `std::variant` but I failed to see the scenario you proposed (probably due to the lack of getters, but I don't recall for sure). Now done via https://github.com/rapidsai/ucxx/pull/121/commits/5037e8fd31a0edce8be8f204619955566c3001ef.